xdsl/irdl/constraints.py

```diff
@@ -20,6 +20,9 @@ class ConstraintContext:
     variables: dict[str, Attribute] = field(default_factory=dict)
     """The assignment of constraint variables."""
 
+    ranges: dict[str, tuple[Attribute, ...]] = field(default_factory=dict)
+    """The assignment of constraint range variables."""
```

Comment:
Should we somehow check somewhere that the variable and range variables have different names?
Otherwise I think the `can_infer` and `infer` funcitons might be incorrect if someone have twice the same name for range variable and variable.

If it's too annoying to check, jsut make an issue for it, as I don't think it's actually gonna happen in practice.

Reply:
I've put this in as the last commit, can revert it if it is deemed too messy